DistroLauncher/WinOobeStrategy.cpp

```diff
@@ -0,0 +1,318 @@
+#include "stdafx.h"
+#include "ApplicationStrategyCommon.h"
+#include "WinOobeStrategy.h"
+
+namespace Oobe
+{
+
+    std::filesystem::path getOobeExePath()
+    {
+        return Win32Utils::thisAppRootdir() / L"ubuntu_wsl_setup.exe";
+    }
+
+    std::filesystem::path getWslConfigPath()
+    {
+        return Win32Utils::homedir() / L".wslconfig";
+    }
+
+    std::wstring createEventName(const wchar_t* const suffix)
+    {
+        return L"Local\\" + DistributionInfo::Name + L'-' + suffix;
+    }
+
+    // Checks the server status file for a successful exit and perform the required post-exit actions.
+    HRESULT serverExitStatusCheck()
+    {
+        if (internal::WslLaunchSuccess(L"grep -E 'DONE|EXITED' /run/subiquity/server-state")) {
+            ExitStatusHandling();
+            return S_OK;
+        }
+        return E_FAIL;
+    }
+
+    // Concatenates [args...] with the common base CLI to be able to run the OOBE in live run mode.
+    // [args...] must contain spaces and quotes and any other metacharacters required to build the command line.
+    // Sample result:
+    // ubuntu_wsl_setup.exe LR"( --no-dry-run --distro-name="Ubuntu-Preview" ) <args...>"
+    //                         |--------------- common cli ------------------|
+    template <typename... Streamable> std::wstring makeCli(Streamable&&... args)
+    {
+        const wchar_t baseCli[] = L" --no-dry-run --distro-name=";
+        return concat(baseCli, std::quoted(DistributionInfo::Name), L' ', std::forward<Streamable>(args)...);
+    }
+
+    // Detects what UI is supported and runs the appropriate callable supplied.
+    // That enables using the same mechanism to run the setup and reconfiguration workflows,
+    // since the requirements for running the OOBE won't change between them.
+    template <typename OnGui, typename OnTui, typename OnNone>
+    HRESULT run_on_autodetect_ui(OnGui onGui, OnTui onTui, OnNone onNone)
+    {
+        static_assert(std::is_convertible_v<OnGui, std::function<HRESULT()>>,
+                      "onGui callback must be a callable returning HRESULT with no arguments");
+        static_assert(std::is_convertible_v<OnTui, std::function<HRESULT()>>,
+                      "onTui callback must be a callable returning HRESULT with no arguments");
+        static_assert(std::is_convertible_v<OnNone, std::function<HRESULT()>>,
+                      "onNone callback must be a callable returning HRESULT with no arguments");
```

Comment:
C++20's concepts would be nice here. Some day 😄.

Reply:
I'd love to jump straight into C++23 to use `std::unreachable`, `std::expected` and more `constexpr`, besides C++20's `std::source_location`, `std::format`, and my favourite: C19 struct designated initialization. C++ source code will look very different after those.